### PR TITLE
add dens.squared

### PR DESCRIPTION
**Pull request by RaphaelS1**

### CODE REVIEW

**Comment by fkiraly** [comment]
@RaphaelS1, @aintoha, why is this doing a massive, inelegant dispatch over existing distributions? This is not very extensible. The "natural" way would be to decorate the distribution with the squared two norm, and then just get it, in one line?

File: R/MeasureDensSquaredloss.R [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      #squarednorm =  distr6::UniformKernel$new()$pdf(0.1)
+
+      pdf = prediction$pdf
+
+      if (is.null(bw)) {
+        score = NA
+      } else {
+        score = -2 * mean(pdf) + mean(squarednorm)/(length(data)^2 * bw)
+      }
+
```
Comment (fkiraly) [inline R/MeasureDensSquaredloss.R:106]: ... something along the lines of
```
pdfeval = prediction$pdf(data)
pdfsquarednorm = prediction$intpdfsquared(-Inf, Inf)

score = -2 * mean(pdfeval) + mean(pdfsquarednorm)

Comment (RaphaelS1) [reply R/MeasureDensSquaredloss.R]: > pdfsquarednorm = prediction$intpdfsquared(-Inf, Inf)

This conflates the predicted distribution which internally uses a Kernel with the Kernel itself.

File: R/MeasureDensSquaredLoss.R [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                           "Unif" = distr6::UniformKernel$new()$pdfSquared2Norm(x = dat),
+                           "Triw" = distr6::Triweight$new()$pdfSquared2Norm(x = dat),
+                           "Logis" = distr6::LogisticKernel$new()$pdfSquared2Norm(x = dat),
+                           "Quart" = distr6::Quartic$new()$pdfSquared2Norm(x = dat),
+                           "Sigm" = distr6::Sigmoid$new()$pdfSquared2Norm(x = dat),
+                           "Silv" = distr6::Silverman$new()$pdfSquared2Norm(x = dat),
+                           "Tric" = distr6::Tricube$new()$pdfSquared2Norm(x = dat),
+                           "Tri" = distr6::TriangularKernel$new()$pdfSquared2Norm(x = dat),
+                           "Cos" = distr6::Cosine$new()$pdfSquared2Norm(x = dat)
+          )
```
Comment (RaphaelS1) [inline R/MeasureDensSquaredLoss.R:65]: A more elegant way to do this would be:

`getFromNamespace(subset(distr6::listKernels(), ShortName == kernel)$ClassName, "distr6")$new()$pdfSquared2Norm(x = dat)`

Comment (aintoha) [reply R/MeasureDensSquaredLoss.R:65]: As discussed with Franz earlier, I include the `pdfSquared2Norm` as output of the distribution to make the loss more general.  I commit the changes to GitHub. Let me know what you think.

File: R/MeasureDensSquaredLoss.R [hunk 3]
```diff
@@ -39,38 +39,42 @@ MeasureDensSquaredloss = R6::R6Class("MeasureDensSquaredloss",
 
         private = list(
         .eps = numeric(0),
-        .score = function(prediction, learner, task, train_set, ...) {
+        .score = function(prediction, learner, task, train_set) {
```
Comment (RaphaelS1) [inline R/MeasureDensSquaredLoss.R:42]: `...` must be included in this function

File: R/MeasureDensSquaredloss.R [hunk 4]
```diff
@@ -39,38 +39,42 @@ MeasureDensSquaredloss = R6::R6Class("MeasureDensSquaredloss",
...
          # change `c("dens.kde")` to list of compatible learners
-           data = task$truth(train_set)
 
-           bw = learner$train(task)$model$bw
-           kernel = learner$train(task)$model$kernel
 
-          # vectorized the data
-          dat <- sapply(data, function (x, y) ((x - y) / bw), y = data)
+           # bw = learner$train(task)$model$bw
+           # kernel = learner$train(task)$model$kernel
```
Comment (RaphaelS1) [inline R/MeasureDensSquaredloss.R:48]: Is this just temporary code? The learner should never be re-trained in the score

File: R/LearnerDensKDE.R [hunk 5]
```diff
@@ -67,12 +67,19 @@ LearnerDensKDE = R6::R6Class("LearnerDensKDE",
...
 
+      dat <-  sapply(task$truth(), function (x, y) ((x - y) / bw), y = task$truth())
 
-      list(distr = Distribution$new(
-          name = paste(self$param_set$values$kernel, "KDE"),
-          short_name = paste0(self$param_set$values$kernel, "_KDE"),
-          pdf = pdf,
-          type = set6::Reals$new()), bw = bw, kernel = self$param_set$values$kernel)
+      pdfSquared2norm <- getFromNamespace(subset(distr6::listKernels(),
+                     ShortName == self$param_set$values$kernel)$ClassName, "distr6")$new()$pdfSquared2Norm(x = dat) / (length(task$truth())^2 * bw)
```
Comment (RaphaelS1) [inline R/LearnerDensKDE.R:73]: Okay this makes sense to me if the value is always dependent on training data and not new values to make predictions on

Comment (RaphaelS1) [reply R/LearnerDensKDE.R:73]: But `self$param_set$values$kernel` should be replaced with `kernel`

File: R/MeasureDensSquaredloss.R (hunk 4 above)
Comment (aintoha) [reply R/MeasureDensSquaredloss.R:48]: I removed this

File: R/LearnerDensKDE.R (hunk 5 above)
Comment (aintoha) [reply R/LearnerDensKDE.R:73]: Yes, the pdfSqaured2Norm only depends on the training data (observation data) 

but `kernel` output `Epan()`.  `self$param_set$values$kernel` have a default value so it will output the right kernel, right?

Comment (RaphaelS1) [reply R/LearnerDensKDE.R:73]: You're right, ignore me!

File: R/MeasureDensSquaredloss.R (hunk 1 above)
Comment (fkiraly) [reply R/MeasureDensSquaredloss.R:78]: I don´t quite get that - would the loss not only see the distribution and the distribution sees the kernel, but the loss doesn´t directly see the kernel?

File: R/MeasureDensSquaredLoss.R (hunk 2 above)
Comment (fkiraly) [reply R/MeasureDensSquaredLoss.R]: this is still strange - it would still do dispatch?

**Comment by fkiraly** [comment]
Hm, I start thinking there is a design issue here.

I don't think loss funtions (or measures) should do dispatch at all on distribution types or method specific properties.

Why do we need this? I think we should quickly have a design review here.

**Comment by aintoha** [comment]
The `pdfSquared2norm` computes the integral from `-Inf` to `Inf` and it only depends on the training data. I'm not sure if we can call directly from the `prediction`

**Comment by fkiraly** [comment]
>The pdfSquared2norm computes the integral from -Inf to Inf and it only depends on the training data. I'm not sure if we can call directly from the prediction

The prediction contains distributions, and `pdfSquared2norm` is a method of these distributions - to me it seems to imply it's perfectly possible? Or do I misunderstand the design here?

Personally, I'd be advocating to make the integrals methods of the distributions, possibly decorated.

File: R/MeasureDensSquaredloss.R (hunk 1 above)
Comment (RaphaelS1) [reply R/MeasureDensSquaredloss.R:78]: > I don´t quite get that - would the loss not only see the distribution and the distribution sees the kernel, but the loss doesn´t directly see the kernel?

Exactly. The loss only sees the fitted distribution, or the predicted pdf/cdf directly. However the distribution is essentially a black-box, the kernel is called within the pdf/cdf and is not a method of the distribution itself and nor is it a composite that wraps a kernel (we could consider this)

**Comment by RaphaelS1** [comment]
> The prediction contains distributions, and pdfSquared2norm is a method of these distributions - to me it seems to imply it's perfectly possible? Or do I misunderstand the design here?

> Personally, I'd be advocating to make the integrals methods of the distributions, possibly decorated.

I think you're getting the `prediction` and the `learner` mixed up, recall the density learners return predicted pdf/cdf and not distributions, distributions are stored in the fitted `learner`. `pdfsquared2norm` is a method of the kernels that are accessed inside the pdf/cdf functions (hard-coded) and not referenced by distributions. So this is not an easy fix as you're suggesting.
My suggestion is as follows: keep it as it is for now and get it working, in the future we can make it more efficient by adding a 'abstract mixture' wrapper to distr6 which would allow custom `Distribution` objects to have a `wrappedModels` field that could then be accessed internally by pdf/cdf and also externally by users. Hence in this case the fitted distributions would have an extra `wrappedModel` field that would be a kernel, then calling these methods would follow simply.
